File: back-kotlin/src/main/kotlin/me/nicolas/stravastats/domain/business/badges/Badge.kt
package me.nicolas.stravastats.domain.business.badges

import me.nicolas.stravastats.domain.business.strava.StravaActivity

sealed class Badge(
    open val label: String,
) {
    abstract fun check(activities: List<StravaActivity>): Pair<List<StravaActivity>, Boolean>

    override fun toString(): String {
        return label
    }
}

File: back-go/domain/business/badge.go
package business

import "mystravastats/domain/strava"

type BadgeSetEnum string

const (
	GENERAL BadgeSetEnum = "GENERAL"
	FAMOUS  BadgeSetEnum = "FAMOUS"
)

type BadgeCheckResult struct {
	Badge       Badge             `json:"badge"`
	Activities  []strava.Activity `json:"activities"`
	IsCompleted bool              `json:"isCompleted"`
}

type Badge interface {
	Check(activities []strava.Activity) (checkedActivities []strava.Activity, isCompleted bool)
	String() string
}

type BaseBadge struct {
	Label string
}

func (b BaseBadge) String() string {
	return b.Label
}
